### PR TITLE
Code cleaning

### PR DESCRIPTION
**Pull request by cyrraz**
I am having a pass through the code to clean and improve it. This PR is the result of a pass through `plotters.py` and `hep_plotters.py`.

It includes a complete externalization of the uncertainty propagation.

### CODE REVIEW

File: plothist/comparison.py [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+    comparison_uncertainties_high = comparison_uncertainties_low
+
+    return (
+        comparison_values,
+        comparison_uncertainties_low,
+        comparison_uncertainties_high,
+    )
+
+
+def compute_difference(hist_1, hist_2, poisson_for_hist_1=False):
```
Comment (0ctagon) [inline plothist/comparison.py:164]: Same comment as for `pull`. A bit strange that the function is not symmetrical for h1 and h2. Why this poisson_for_hist_1 argument? why not also a poisson_for_hist_2? We could use kwargs.

File: plothist/comparison.py [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+    Compute the difference between two histograms.
+
+    Parameters
+    ----------
+    hist_1 : boost_histogram.Histogram
+        The first histogram.
+    hist_2 : boost_histogram.Histogram
+        The second histogram.
+    poisson_for_hist_1 : bool, optional
+        Whether to use Poisson uncertainties for hist_1. Default is False.
```
Comment (0ctagon) [inline plothist/comparison.py:175]: Add info: `Default is False -> Default uses the square root of the sum of the variances.`

File: plothist/comparison.py [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+        comparison_uncertainties_high = comparison_uncertainties_low
+
+    return (
+        comparison_values,
+        comparison_uncertainties_low,
+        comparison_uncertainties_high,
+    )
+
+
+def compute_ratio(
```
Comment (0ctagon) [inline plothist/comparison.py:208]: Same comment as pull or diff for the poisson for h1.

File: plothist/comparison.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+    Parameters
+    ----------
+    hist_1 : boost_histogram.Histogram
+        The numerator histogram.
+    hist_2 : boost_histogram.Histogram
+        The denominator histogram.
+    ratio_uncertainty : str, optional
+        How to treat the uncertainties of the histograms: "uncorrelated" for simple comparison, "split" for scaling and split hist_1 and hist_2 uncertainties. Default is "uncorrelated".
+    poisson_for_hist_1 : bool, optional
+        Whether to use Poisson uncertainties for hist_1. Default is False.
```
Comment (0ctagon) [inline plothist/comparison.py:223]: Default is sqrt

File: plothist/comparison.py [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+        raise ValueError("ratio_uncertainty not in ['uncorrelated', 'split'].")
+
+    return (
+        comparison_values,
+        comparison_uncertainties_low,
+        comparison_uncertainties_high,
+    )
+
+
+def compute_comparison(
```
Comment (0ctagon) [inline plothist/comparison.py:281]: Same comment as above for h1 poisson.

File: plothist/hep_plotters.py
Comment (0ctagon) [inline plothist/hep_plotters.py:1]: The plot MC stat unc part should be a function.

File: plothist/plotters.py [hunk 6]
```diff
@@ -598,6 +551,7 @@ def plot_comparison(
     comparison_ylabel=None,
     comparison_ylim=None,
     ratio_uncertainty="uncorrelated",
+    poisson_for_hist_1=False,
```
Comment (0ctagon) [inline plothist/plotters.py:554]: Same comment.

File: plothist/hep_plotters.py [hunk 7]
```diff
@@ -130,76 +136,20 @@ def compare_data_mc(
...
     if not mc_uncertainty:
-        mc_hist_total[:] = np.stack(
-            [mc_hist_total.values(), np.zeros_like(mc_hist_total.values())], axis=-1
-        )
+        mc_hist_total[:] = np.c_[
+            mc_hist_total.values(), np.zeros_like(mc_hist_total.values())
+        ]
 
     # Compute data uncertainties
     if np.allclose(data_hist.variances(), data_hist.values()):
```
Comment (0ctagon) [inline plothist/hep_plotters.py:144]: Use `_is_unweighted()`

File: plothist/plotters.py [hunk 8]
```diff
@@ -598,6 +551,7 @@ def plot_comparison(
     comparison_ylabel=None,
     comparison_ylim=None,
     ratio_uncertainty="uncorrelated",
```
Comment (0ctagon) [inline plothist/plotters.py:553]: Add this to a `kwargs`? I find it strange to have it as an argument if it only impact 2/4 comparison method.

File: plothist/plotters.py [hunk 9]
```diff
@@ -719,6 +639,13 @@ def plot_comparison(
             ax.set_ylabel(r"$\frac{" + h1_label + "}{" + h2_label + "}$")
 
         if ratio_uncertainty == "split":
+            np.seterr(divide="ignore", invalid="ignore")
+            h2_scaled_uncertainties = np.where(
```
Comment (0ctagon) [inline plothist/plotters.py:643]: Why an uncertainty calculation here? This should be taken care of by `compute_comparison()`

File: plothist/comparison.py [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+
+    Raise
+    -----
+    ValueError
+        If the histogram is weighted.
+
+    """
+    if not _is_unweighted(data_hist):
+        raise ValueError(
+            "Poisson uncertainties make sense only for unweighted histograms."
```
Comment (0ctagon) [inline plothist/comparison.py:46]: After a lot of discussions, this sentence is not unanimous. We should either trust the user to use poisson correctly, or put documentation on why it is not possible.

File: plothist/comparison.py [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+
+    if comparison == "ratio":
+        values, lower_uncertainties, upper_uncertainties = compute_ratio(
+            hist_1, hist_2, ratio_uncertainty, poisson_for_hist_1
+        )
+    elif comparison == "relative_difference":
+        values, lower_uncertainties, upper_uncertainties = compute_ratio(
+            hist_1, hist_2, ratio_uncertainty, poisson_for_hist_1
+        )
+        values -= 1  # relative difference is ratio-1
```
Comment (0ctagon) [inline plothist/comparison.py:326]: beautiful.

**Review by 0ctagon — CHANGES_REQUESTED** [review]
Beautiful PR, we are getting closer and closer to v1.0

File: plothist/comparison.py (hunk 10 above)
Comment (cyrraz) [reply plothist/comparison.py:46]: I am not saying it is impossible (I am not sure), but I do not know how to compute the Poisson uncertainties for an unweighted histogram. I changed the sentence.

File: plothist/comparison.py [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+        hist_1.variances() / hist_2.values() ** 2
+        + hist_2.variances() * hist_1.values() ** 2 / hist_2.values() ** 4,
+        np.nan,
+    )
+    np.seterr(divide="warn", invalid="warn")
+
+    return ratio_variances
+
+
+def compute_pull(hist_1, hist_2, poisson_for_hist_1=False):
```
[the remark this answers is not in the document]
Comment (cyrraz) [reply plothist/comparison.py:116]: Poisson for hist_2 is not supported in this version, I changed the name

File: plothist/comparison.py (hunk 2 above)
Comment (cyrraz) [reply plothist/comparison.py:175]: Let's add a stat. section later in the doc

File: plothist/hep_plotters.py
Comment (cyrraz) [reply plothist/hep_plotters.py:1]: Yes, not the purpose of this PR, but let's keep this in mind.

File: plothist/plotters.py (hunk 8 above)
Comment (cyrraz) [reply plothist/plotters.py:553]: discussed in meeting

File: plothist/plotters.py (hunk 9 above)
Comment (cyrraz) [reply plothist/plotters.py:643]: discussed in meeting

File: plothist/comparison.py (hunk 2 above)
Comment (0ctagon) [reply plothist/comparison.py]: Just after this PR

File: plothist/hep_plotters.py
Comment (0ctagon) [reply plothist/hep_plotters.py:1]: I think it enters the "code cleaning" part of this PR.

File: plothist/comparison.py (hunk 10 above)
Comment (0ctagon) [reply plothist/comparison.py]: Since we don't know if it's possible, shouldn't we just raise a `Warning` except of a `ValueError`?

**Pull request merged**